algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+          have h := classical.some_spec (subring.one_mem $ is_hom.im f),
+          simp [is_hom.map_add f, is_hom.map_neg f, h, is_hom.map_one f]
+        end },
+  hfg := λ ⟨x, hx⟩, subtype.eq (by simp [first_isom._match_1]; simpa using classical.some_spec hx),
+  hgf :=
+    begin
+      intro x,
+      cases is_ideal.coset_rep x with y hy,
+      rw ← hy,
+      simp [first_isom._match_1],
+      have hz := @classical.some_spec _ (λ z, f z = f y) ⟨y, rfl⟩,
+      simp [is_hom.map_add f, hz, is_hom.map_neg f]
+    end
+}
+
+local infix `^` := monoid.pow
+
+def nilpotents (α : Type u) [comm_ring α] : set α := { x | ∃ n, x^(nat.succ n) = 0 }
+def is_unit {α : Type u} [comm_ring α] (x : α) := ∃ y, x * y = 1
+def nonunits (α : Type u) [comm_ring α] : set α := { x | ¬∃ y, x * y = 1 }
+
+
+-- page 3
+
+section principal_ideal
+
+variables {α : Type u} [comm_ring α] (x : α)
+
+def principal_ideal : set α := { y | ∃ z, x * z = y }
+
+instance principal_ideal.is_ideal : is_ideal α $ principal_ideal x :=
+{ zero_mem := ⟨0, mul_zero x⟩,
+  add_mem  := λ y₁ y₂ ⟨z₁, hz₁⟩ ⟨z₂, hz₂⟩, ⟨z₁ + z₂, by simp [left_distrib, hz₁, hz₂]⟩,
+  mul_mem  := λ y₁ y₂ ⟨z₁, hz₁⟩, ⟨z₁ * y₂, by simp [hz₁.symm, mul_assoc]⟩ }
+
+variable (α)
+
+theorem principal_ideal_one_eq_univ : principal_ideal (1:α) = univ_ideal α :=
+set.ext $ λ x, ⟨λ hx, trivial, λ hx, ⟨x, one_mul x⟩⟩
+
+variable {α}
+
+@[simp] lemma mem_principal_ideal : x ∈ principal_ideal x :=
+⟨1, mul_one x⟩
+
+theorem unit_iff_principal_ideal_eq_one : is_unit x ↔ principal_ideal x = principal_ideal 1 :=
+⟨λ ⟨y, hy⟩, set.ext $ λ z, ⟨λ hz, ⟨z, one_mul z⟩, λ hz, ⟨y * z, by rw [← mul_assoc, hy, one_mul]⟩⟩,
+ λ hx, show (1:α) ∈ principal_ideal x, by simp [hx]⟩
+
+variable {x}
+
+theorem principal_ideal.subset_of_mem {S : set α} [is_ideal α S] : x ∈ S → principal_ideal x ⊆ S :=
+λ h z ⟨w, hw⟩, set.mem_of_eq_of_mem hw.symm $ is_ideal.mul_mem h
+
+variable (α)
+
+theorem principal_ideal_zero_eq_zero_ideal : principal_ideal (0:α) = zero_ideal α :=
+set.ext $ λ x, ⟨λ ⟨y, hy⟩, by rw [← hy]; simpa [zero_mul], λ hx, ⟨0, by rw [eq_of_mem_singleton hx, zero_mul]⟩⟩
+
+end principal_ideal
+
+theorem is_ideal.eq_univ_of_contains_unit {α : Type u} [comm_ring α] (S : set α) [is_ideal α S] :
+(∃ x:α, x ∈ S ∧ is_unit x) → S = set.univ :=
+λ ⟨x, hx, y, hy⟩, set.ext $ λ z, ⟨λ hz, trivial, λ hz, calc
+   z = (x * y) * z : by simp [hy]
+ ... = x * (y * z) : mul_assoc x y z
+ ... ∈ S : is_ideal.mul_mem hx⟩
+
+theorem is_ideal.univ_of_one_mem {α : Type u} [comm_ring α] (S : set α) [is_ideal α S] :
+(1:α) ∈ S → S = set.univ :=
+λ h, set.ext $ λ z, ⟨λ hz, trivial, λ hz, by simpa using (is_ideal.mul_mem h : 1 * z ∈ S)⟩
+
+
+-- Proposition 1.2 start
+
+section prop_1_2
+
+variables (α : Type u) [comm_ring α] (zero_ne_one : (0:α) ≠ 1)
+
+class is_field : Prop :=
+(h : ∀ {x:α}, x ≠ 0 → is_unit x)
+(zero_ne_one : (0:α) ≠ 1)
+
+class ideal_eq_zero_or_univ : Prop :=
+(h : ∀ (S : set α) [is_ideal α S], S = zero_ideal α ∨ S = univ_ideal α)
+
```

Comment:
Why should this be a class? Why do I even care about it? It's just some technical thing. I care about whether or not something is a field enough to make is_field a class, but this?